OCPBUGS-43344: OCPBUGS-42244: Implement namespaceSelectors on ingress with all pods IPs

pliurh opened the pull request:
During SDN live migration, the VNID from ovn-k nodes cannot be used distinguish namespaces. So generate per-pod-IP match rules instead.

pliurh commented:
/retest-required

pliurh commented:
/retest-required

pliurh commented:
/retest-required

pliurh commented:
/retest-required

pliurh commented:
@danwinship PTAL

pliurh commented:
/jira refresh

danwinship commented on pkg/network/node/networkpolicy.go:
```diff
@@ -789,15 +769,15 @@ func (np *networkPolicyPlugin) parsePeerFlows(npns *npNamespace, npp *npPolicy,
-				// We can implement namespaceSelectors on ingress by just
-				// checking the source VNID...
-				npp.watchesNamespaces = true
-				peerFlows = append(peerFlows, np.selectNamespaces(peer.NamespaceSelector, dir)...)
```
Do we not have some way of distinguishing that we are in "migration mode"?

It's _much_ less efficient to use per-IP matches rather than VNID matches. This will break some very large clusters.

martinkennelly replied:
@danwinship even if we can detect migration and only implement this per IP matches during that time period, do you also foresee issues with large clusters? 
I guess so..

pliurh replied:
> Do we not have some way of distinguishing that we are in "migration mode"?

No. We could add an environment variable for that if needed.

But currently per-IP is also used by egress policies, would it get this scale problem?

danwinship replied:
> even if we can detect migration and only implement this per IP matches during that time period, do you also foresee issues with large clusters?

Yes. In theory, some large clusters may have NP-related performance problems during migration that they did not have before migration.

I don't have any good estimates of how large a cluster needs to be to hit this problem though; SDN has always had this particular optimization, so we don't have any evidence of how different customers' clusters would behave without it...

> But currently per-IP is also used by egress policies, would it get this scale problem?

Yes, but egress policies have always worked that way, so people would have already avoided creating egress policies that caused problems in their cluster.

pliurh replied:
I updated the patch. Now, the per-podIP rules will be added after the VNI rules and only when running in migration mode. Also, I take the host-network namespace into account. PTAL

pliurh committed:
Implement namespaceSelectors on ingress with all pods IPs

During SDN live migration, the VNID from ovn-k nodes cannot be
used distinguish namespaces. So generate per-pod-IP match rules
instead.

For host network namespace, we need to rules with the OVN mp0 IPs
for the host to pod traffic from the ovn-k nodes.

Signed-off-by: Peng Liu <pliu@redhat.com>

pliurh commented:
Relies on https://github.com/openshift/cluster-network-operator/pull/2532

pliurh commented:
/retest-required

danwinship commented on pkg/network/node/networkpolicy.go:
```diff
@@ -35,6 +37,8 @@ const HostNetworkNamespace = "openshift-host-network"
 // only used by network policy functions
 const NetPolIsolationCookie = 1147582955
 
+const migrtionEnvVar = "NODE_CNI"
```
```suggestion
const migrationEnvVar = "NODE_CNI"
```

danwinship commented on pkg/network/node/networkpolicy.go:
```diff
@@ -794,6 +798,35 @@ func (np *networkPolicyPlugin) parsePeerFlows(npns *npNamespace, npp *npPolicy,
 				// checking the source VNID...
 				npp.watchesNamespaces = true
 				peerFlows = append(peerFlows, np.selectNamespaces(peer.NamespaceSelector, dir)...)
+				if _, ok := os.LookupEnv(migrtionEnvVar); ok {
```
do the `LookupEnv` in `NewNetworkPolicyPlugin` and stash the result in a field on `networkPolicyPlugin`

danwinship commented on pkg/network/node/networkpolicy.go:
```diff
@@ -794,6 +798,35 @@ func (np *networkPolicyPlugin) parsePeerFlows(npns *npNamespace, npp *npPolicy,
+					npp.watchesAllPods = true
+					peerFlows = append(peerFlows, np.selectPodsFromNamespaces(peer.NamespaceSelector, &metav1.LabelSelector{}, dir)...)
+
+					// If the host network namespace is selected, Add rules for
+					// the OVN mp0 IP of each node.
```
SDN treats all non-pod-cidr traffic as "host network", so we just need to accept anything that's not in `npp.node.clusterCIDRs`.

The "not" makes this difficult though... the easiest way to deal with this would probably be to update [this rule](https://github.com/openshift/sdn/blob/d515d15883394318af2c55e75e17291a65d023b2/pkg/network/node/ovscontroller.go#L258) to add `load:1->NXM_NX_REG3[]` to the actions, to tag all nw_src=clusterCIDR with reg3=1 (doesn't need to be conditional on migration). Then here you can just check `reg3=0` here to match everything that didn't match that (ie, non-clusterCIDR-source traffic).

Oh! And either way, you need to bump `ruleVersion` in ovscontroller.go (which will require updates to the unit tests as well).

pliurh replied:
But the IP of the mp0 interface is in the range of the clusterCIDR. It's the second IP of the node subnet.

pliurh replied:
I added unit test and bumped the `ruleVersion`. PTAL

pliurh commented:
/retest

pliurh commented:
/retest-required

danwinship commented on deps.diff:
```diff
@@ -1,9 +0,0 @@
-diff --no-dereference -N -r current/vendor/github.com/containernetworking/plugins/plugins/ipam/host-local/backend/disk/backend.go updated/vendor/github.com/containernetworking/plugins/plugins/ipam/host-local/backend/disk/backend.go
```
We're carrying a patch against upstream `containernetworking/plugins` (or at least, against the version of that module that we're vendoring). So this file needs to be kept, and you need to manually undo the changes that "go mod vendor" tries to make to `vendor/github.com/containernetworking/plugins/`.

(Maybe I should hack this to work differently because it's constantly a pain to deal with...)

danwinship commented on pkg/network/node/networkpolicy.go:
```diff
@@ -794,6 +802,35 @@ func (np *networkPolicyPlugin) parsePeerFlows(npns *npNamespace, npp *npPolicy,
+					// the OVN mp0 IP of each node.
+					sel, _ := metav1.LabelSelectorAsSelector(peer.NamespaceSelector)
+					if _, ok := np.nsMatchCache[sel.String()].matches[HostNetworkNamespace]; ok {
+						hsns, err := np.node.osdnInformers.Network().V1().HostSubnets().Lister().List(labels.Everything())
```
ok, since this is only for migration mode, and we shouldn't be adding or removing nodes while migrating, you could just precompute the list of mp0-related flows in `Start()` rather than regenerating it every time.

Alternatively... OVS lets you match IPs against _arbitrary_ bitmasks, not just proper CIDR expressions. If `node.networkInfo.ClusterNetworks` is `[{ ClusterCIDR: "10.128.0.0/14", HostSubnetLength: 23 }]`, then the mp0 IPs are exactly the ones that have "`10.128`" in bits 0-14 and "`2`" in bits 23-32 (and anything in bits 15-22). So that's `nw_src=10.128.0.2/0xfffc1f` (where that mask value is `uint32(math.MaxUint32 << (32 - cidrMaskLen)) | (math.MaxUint32 >> hostSubnetLen)`)

Right?

pliurh replied:
Done. PTAL

pliurh commented:
/retest-required

pliurh commented:
/jira refresh

pliurh commented:
/retest-required

danwinship commented:
/lgtm
/approve
/retest-required
/label backport-risk-assessed

asood-rh commented:
/label cherry-pick-approved

pliurh commented:
/jira refresh

pliurh commented:
/cherry-pick release-4.15
